dubbo-remoting/dubbo-remoting-zookeeper/src/main/java/org/apache/dubbo/remoting/zookeeper/curator/CuratorZookeeperClient.java

```diff
@@ -218,47 +218,48 @@ public void doClose() {
     protected CuratorZookeeperClient.CuratorWatcherImpl createTargetDataListener(String path, DataListener listener) {
-        return new CuratorWatcherImpl(client, listener);
+        return new CuratorWatcherImpl(client, listener, path);
     }
 
     @Override
-    protected void addTargetDataListener(String path, CuratorZookeeperClient.CuratorWatcherImpl treeCacheListener) {
-        this.addTargetDataListener(path, treeCacheListener, null);
+    protected void addTargetDataListener(String path, CuratorZookeeperClient.CuratorWatcherImpl nodeCacheListener) {
+        this.addTargetDataListener(path, nodeCacheListener, null);
     }
 
     @Override
-    protected void addTargetDataListener(String path, CuratorZookeeperClient.CuratorWatcherImpl treeCacheListener, Executor executor) {
+    protected void addTargetDataListener(String path, CuratorZookeeperClient.CuratorWatcherImpl nodeCacheListener, Executor executor) {
         try {
-            TreeCache treeCache = TreeCache.newBuilder(client, path).setCacheData(false).build();
-            treeCacheMap.putIfAbsent(path, treeCache);
-
+            NodeCache nodeCache = new NodeCache(client, path);
+            if (nodeCacheMap.putIfAbsent(path, nodeCache) != null) {
+                return;
+            }
             if (executor == null) {
-                treeCache.getListenable().addListener(treeCacheListener);
+                nodeCache.getListenable().addListener(nodeCacheListener);
             } else {
-                treeCache.getListenable().addListener(treeCacheListener, executor);
+                nodeCache.getListenable().addListener(nodeCacheListener, executor);
             }
 
-            treeCache.start();
+            nodeCache.start();
         } catch (Exception e) {
-            throw new IllegalStateException("Add treeCache listener for path:" + path, e);
+            throw new IllegalStateException("Add nodeCache listener for path:" + path, e);
         }
     }
 
     @Override
-    protected void removeTargetDataListener(String path, CuratorZookeeperClient.CuratorWatcherImpl treeCacheListener) {
-        TreeCache treeCache = treeCacheMap.get(path);
-        if (treeCache != null) {
-            treeCache.getListenable().removeListener(treeCacheListener);
+    protected void removeTargetDataListener(String path, CuratorZookeeperClient.CuratorWatcherImpl nodeCacheListener) {
+        NodeCache nodeCache = nodeCacheMap.get(path);
+        if (nodeCache != null) {
+            nodeCache.getListenable().removeListener(nodeCacheListener);
         }
-        treeCacheListener.dataListener = null;
+        nodeCacheListener.dataListener = null;
     }
 
     @Override
     public void removeTargetChildListener(String path, CuratorWatcherImpl listener) {
         listener.unwatch();
     }
 
-    static class CuratorWatcherImpl implements CuratorWatcher, TreeCacheListener {
+    static class CuratorWatcherImpl implements CuratorWatcher, NodeCacheListener {

```

Comment:
I suggest we separate CuratorWatcherImpl into two classes, one with a single responsbility.

Reply:
agree